### PR TITLE
Permissions setter

### PR DESCRIPTION
**Pull request by MorganRodgers**
Addresses requirement from #11.

### CODE REVIEW

**Comment by ericfranz** [comment]
So the issue I wrote up was poorly written.


1. In regards to the mapping file, it is created here:

https://github.com/OSC/shiny_admin/blob/7ab9e1f9ff84f628818817d009493e7e1ae2426e/app/models/mapping.rb#L39-L41

If the permissions are 664 updating a mappings file preserve the permissions, or should we set permissions again after writing the file?

2. The ~bmiadmin/share/ directory will need to be 775 so that users in that group can create files underneath. That is probably a setup step that needs to take place. Perhaps the initializer should raise an exception if this is not the case, since and admin would not be able to create files underneath then (where an admin is in the group of users that have write access to this directory, since we no longer have the bmiadmin user accessing it)

3. So the database is created in `bin/setup-production` (is this still what we want)? So if an admin user (not bmiadmin) creates user accessing the app, then we want the permissions on the database to be 664. Presumably this is what will happen since the `bin/setup-production` is executed, the database will be created when the app is launched from the dashboard and then the admin app is loaded, which starts the initializer and runs this code. 

Should we avoid touching the database file, since the point seems to be to create it if it doesn't exist? What happens if an empty file is created there and then `bin/setup-production` is run after the fact?

**Comment by MorganRodgers** [comment]
A little hard to follow some of your sentences, but here goes:

1. Permissions are preserved across overwrites, so long as the file is not deleted first, so this is safe from the application's stand point, but not against user actions.
2. For this to work all intermediate directories including $HOME need to be set to mode 755. I wouldn't normally want to change a user's $HOME permissions in the application, but since this is a shared service account it should be fine. I will add permissions setting for directories as well.
3. I don't have a preference where we initialize the database.

Unnumbered: What happens is that the select will fail and application setup will fail. That said it looks to me like setup always runs before the initializers?

**Comment by ericfranz** [comment]
> so long as the file is not deleted first

That was my only concern.

> I will add permissions setting for directories as well.

I think that for the bmiadmin user this can be done once. Other users in PAS1203 wouldn't be able to fix that.

File: config/initializers/permissions_setter.rb [hunk 1]
```diff
@@ -0,0 +1,13 @@
+# Be sure to restart your server when you modify this file.
+
+# You can add backtrace silencers for libraries that you're using but don't wish to see in your backtraces.
+# Rails.backtrace_cleaner.add_silencer { |line| line =~ /my_noisy_library/ }
+
+# You can also remove all the silencers if you're trying to debug a problem that might stem from framework code.
+# Rails.backtrace_cleaner.remove_silencers!
+require 'fileutils'
+
+FileUtils.touch(Configuration.yaml_file_path)
```
Comment (ericfranz) [inline config/initializers/permissions_setter.rb]: Before touching files, do

```ruby
File.umask(2)
```

Since all of the admin users of this Rails app will not be the bmiadmin user.

Comment (ericfranz) [reply config/initializers/permissions_setter.rb]: Also `FileUtils.chmod(0775 ` should be `FileUtils.chmod(0664`. That was my bad on the issue.

**Comment by ericfranz** [comment]
Oh yeah setup always runs before the initializers as long as the app is launched through the dashboard. If you access the URL directly you get a different outcome.

**Comment by MorganRodgers** [comment]
I added a warning to the Mapping index action. It is not dismissable, but does contain the command to remediate the problem. I think that will be better than aborting the application partway through setup, or startup.

I have removed the touch and initializer in favor of re-setting the permissions on `mappings.yaml` and `mappings.sqlite3` when a new mapping is saved, or one is deleted.

File: app/views/mappings/index.html.erb [hunk 2]
```diff
@@ -1,3 +1,9 @@
+<% if Mapping.has_directory_permission_errors? %>
+  <div class="alert alert-danger" role="alert">
+    <div class="panel-body">Warning: all directories between <code><%= Configuration.home %></code> and <code><%= Configuration.production_database_path %></code> must have their permissions set to <code>0775</code>. Run the following as user <code>bmiadmin</code>:</div>
```
Comment (ericfranz) [inline app/views/mappings/index.html.erb]: I'd remove `Configuration.home` and just say "all directories above #{Configuration.production_database_path}`. Reason is that `Configuration.home` is set to `Pathname.new(ENV['HOME'])` (which is also `Dir.home`) which is incorrect. The database is stored in bmiadmin's home directory, but the PUN process will be each admin user that is in PAS1203 group, not bmiadmin.

So if `efranz` is running the Shiny admin app, `Configuration.home` would be /users/PZS0562/efranz and `Configuration.production_database_path` would be /users/PAS1203/bmiadmin/awesim/share/mappings.sqlite3

That might make ascend preferable, because you don't have to embed `Dir.home('bmiadmin')` in the app.

Or you could do just that: make `Configuration.home` the value of `Dir.home('bmiadmin')`. I think you are setting the ADMIN_USER somewhere (and ADMIN_GROUP?).

Another observation...

```
efranz@owens-login02:/users/PAS1203/bmiadmin/awesim/share$ stat -c "%G" . .. ../.. ../../.. ../../../..
PAS1203
PAS1203
PAS1203
root
root
```

Comment (ericfranz) [reply app/views/mappings/index.html.erb]: That last observation was: could ascend and check permissions if the group is the ADMIN_GROUP. Having the admin group and the admin user known by the app is the "same difference" though... Is there any problem with just checking up the directory tree, regardless of whether the admin user or root owns the directory?

**Comment by ericfranz** [comment]
I don’t see my code comment so I must have not clicked submit. The method permission_sensitive_dirs results in infinit loop or raises exception if Pathname.new("/").parent raises exception when $HOME is not in the parent tree of the database, which is now the case.

You can avoid the while loop altogether if you use ascend and exclude dirs owned by root; and that avoids the need to have a properly set `Configuration.home`

File: app/views/mappings/index.html.erb [hunk 3]
```diff
@@ -11,7 +17,7 @@
     <th></th>
   </thead>
   <tbody>
-  <%=  Mapping.find_each do |mapping| %>
+  <%  Mapping.find_each do |mapping| %>
```
Comment (ericfranz) [inline app/views/mappings/index.html.erb:20]: In Rails, for things like this its best practice to do the query in the controller and set an instance variable like `@mappings` that the view uses.

This slightly decouples the view from the model slightly has these benefits:

1. Maybe makes automated testing easier (for example, I can write controller tests and all the logic is in the controller)
2. It is easier to modify the logic e.g.  vary the `@mappings` list queried based on a query param
3. Could reuse the view code from multiple controllers

There are probably other reasons too but those are the only ones I remember on the top of my head.

File: config/configuration_singleton.rb [hunk 4]
```diff
@@ -46,6 +46,10 @@ def load_dotenv_files
     Dotenv.load(*dotenv_files)
   end
 
+  def production?
+    ENV['RAILS_ENV'] == 'production'
```
Comment (ericfranz) [inline config/configuration_singleton.rb:50]: This should be:

```
rails_env == 'production'
```

See private method at bottom.

File: app/views/mappings/index.html.erb [hunk 5]
```diff
@@ -1,3 +1,9 @@
+<% if Mapping.has_directory_permission_errors? %>
+  <div class="alert alert-danger" role="alert">
+    <div class="panel-body">Warning: certain directories owned by <code>bmiadmin</code> have incorrect permissions which will prevent app and dataset sharing.. As <code>bmiadmin</code> run the following to set the correct permissions:</div>
+    <div class="panel-body"><pre><code><%= Mapping.directory_permissions_command %></code></pre></div>
+  </div>
+<% end %>
```
Comment (ericfranz) [inline app/views/mappings/index.html.erb:6]: We already have code https://github.com/OSC/shiny_admin/blob/db773834b39496aae97b81685503709d1bd0bf2c/app/views/layouts/application.html.erb#L55-L63 to handle displaying warning messages; the Rails way wouuld be to add a warning message to the flash object in the controller. Normally the flash object is so the error message appears after a redirect, but you can use https://guides.rubyonrails.org/action_controller_overview.html#flash-now to have it set when rendering. You won't get the exact look of this code, but it will be close:

```ruby
flash.now[:error] = "Warning: certain directories owned by <code>bmiadmin</code> have incorrect permissions which will prevent app and dataset sharing.. As <code>bmiadmin</code> run the following to set the correct permissions: <pre><code><%= Mapping.directory_permissions_command %></code></pre>".html_safe
```
